<a href="https://colab.research.google.com/github/Bharat-Moorthy/MLAI---Paathsalaassignments/blob/main/BBC_ML_model_with_TFIDF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [26]:
import pandas as pd

Inputting the dataset

In [27]:
news=pd.read_csv("/content/BBC News.csv")

In [28]:
news

ArticleId                                               Text  \
0          1833  worldcom ex-boss launches defence lawyers defe...   
1           154  german business confidence slides german busin...   
2          1101  bbc poll indicates economic gloom citizens in ...   
3          1976  lifestyle  governs mobile choice  faster  bett...   
4           917  enron bosses in $168m payout eighteen former e...   
...         ...                                                ...   
1485        857  double eviction from big brother model caprice...   
1486        325  dj double act revamp chart show dj duo jk and ...   
1487       1590  weak dollar hits reuters revenues at media gro...   
1488       1587  apple ipod family expands market apple has exp...   
1489        538  santy worm makes unwelcome visit thousands of ...   

           Category  
0          business  
1          business  
2          business  
3              tech  
4          business  
...             ...  
1485  entertainment  
1486  entertainment  
1487       business  
1488           tech  
1489           tech  

[1490 rows x 3 columns]

The unqiue values in the target variable are 

In [29]:
news["Category"].unique()

array(['business', 'tech', 'politics', 'sport', 'entertainment'],
      dtype=object)

In [30]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [31]:
tfidf=TfidfVectorizer(sublinear_tf=True, min_df=5, norm='l2', encoding='latin-1', ngram_range=(1, 2), stop_words='english')
tfidf_transform=tfidf.fit_transform(news["Text"])

Split the dataset into X and Y

In [32]:
x=pd.DataFrame(tfidf_transform.toarray(), columns=tfidf.get_feature_names_out())

In [33]:
y=news["Category"]

We have to do the label encoding for the target variable in order to proceed for modelling 

In [34]:
from sklearn.preprocessing import LabelEncoder

In [35]:
# initialize LabelEncoder
lencod = LabelEncoder()
# fit_transform() converts the text to numbers
news['Category'] = lencod.fit_transform(news['Category'])

In [36]:
y=news['Category']

In [37]:
y

0       0
1       0
2       0
3       4
4       0
       ..
1485    1
1486    1
1487    0
1488    4
1489    4
Name: Category, Length: 1490, dtype: int64

Splitting the dataset into train and test 

In [38]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size = 0.25, random_state = 42)

In [39]:
X_train

000  000 complaints  000 copies  000 fans  000 fine  000 jobs  \
179   0.000000             0.0         0.0       0.0       0.0       0.0   
1439  0.000000             0.0         0.0       0.0       0.0       0.0   
1237  0.000000             0.0         0.0       0.0       0.0       0.0   
841   0.000000             0.0         0.0       0.0       0.0       0.0   
1462  0.000000             0.0         0.0       0.0       0.0       0.0   
...        ...             ...         ...       ...       ...       ...   
1130  0.000000             0.0         0.0       0.0       0.0       0.0   
1294  0.000000             0.0         0.0       0.0       0.0       0.0   
860   0.044139             0.0         0.0       0.0       0.0       0.0   
1459  0.047149             0.0         0.0       0.0       0.0       0.0   
1126  0.000000             0.0         0.0       0.0       0.0       0.0   

      000 new  000 people  000 workers  000 year  ...  zach  zach braff  \
179       0.0    0.000000          0.0       0.0  ...   0.0         0.0   
1439      0.0    0.000000          0.0       0.0  ...   0.0         0.0   
1237      0.0    0.000000          0.0       0.0  ...   0.0         0.0   
841       0.0    0.000000          0.0       0.0  ...   0.0         0.0   
1462      0.0    0.000000          0.0       0.0  ...   0.0         0.0   
...       ...         ...          ...       ...  ...   ...         ...   
1130      0.0    0.000000          0.0       0.0  ...   0.0         0.0   
1294      0.0    0.000000          0.0       0.0  ...   0.0         0.0   
860       0.0    0.000000          0.0       0.0  ...   0.0         0.0   
1459      0.0    0.049437          0.0       0.0  ...   0.0         0.0   
1126      0.0    0.000000          0.0       0.0  ...   0.0         0.0   

      zealand  zeppelin  zero  zero tolerance  zhang  zone  zurich  \
179       0.0       0.0   0.0             0.0    0.0   0.0     0.0   
1439      0.0       0.0   0.0             0.0    0.0   0.0     0.0   
1237      0.0       0.0   0.0             0.0    0.0   0.0     0.0   
841       0.0       0.0   0.0             0.0    0.0   0.0     0.0   
1462      0.0       0.0   0.0             0.0    0.0   0.0     0.0   
...       ...       ...   ...             ...    ...   ...     ...   
1130      0.0       0.0   0.0             0.0    0.0   0.0     0.0   
1294      0.0       0.0   0.0             0.0    0.0   0.0     0.0   
860       0.0       0.0   0.0             0.0    0.0   0.0     0.0   
1459      0.0       0.0   0.0             0.0    0.0   0.0     0.0   
1126      0.0       0.0   0.0             0.0    0.0   0.0     0.0   

      zurich premiership  
179                  0.0  
1439                 0.0  
1237                 0.0  
841                  0.0  
1462                 0.0  
...                  ...  
1130                 0.0  
1294                 0.0  
860                  0.0  
1459                 0.0  
1126                 0.0  

[1117 rows x 9927 columns]

Now we will do the Modelling using different models and find out which one gives the best results

In [42]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix, classification_report

In [41]:
import time 
from tqdm import tqdm
from sklearn.model_selection import cross_val_score
def select_model(X, Y):

    best_models = {}
    model = [
        {   'name': 'LogisticRegression',
            'estimator': LogisticRegression(),
         
        },
        {   'name': 'RandomForestClassifier',
            'estimator': RandomForestClassifier(),
         
        },
        {   'name': 'GradientBoostingRegressor',
            'estimator': GradientBoostingClassifier(),
         
        },
        {   'name': 'XGBoost',
            'estimator': XGBClassifier(),
          
        },
        {   'name': 'BaggingClassifier',
            'estimator': BaggingClassifier(),
          
        },
        {   'name': 'MultinomialNB',
            'estimator': MultinomialNB()
              
        }
        
    ]

    for model in tqdm(model):
        start = time.time()
        cv_results=cross_val_score(model["estimator"],X,y,cv=5,scoring="accuracy")
        best_models[model['name']] = {'score': cv_results.mean(),'model':model['estimator']}
        end=time.time()
        run = end - start
        
    return best_models

model = select_model(x,y)
model

100%|██████████| 6/6 [21:05<00:00, 210.98s/it]


{'LogisticRegression': {'score': 0.9751677852348992,
  'model': LogisticRegression()},
 'RandomForestClassifier': {'score': 0.9516778523489933,
  'model': RandomForestClassifier()},
 'GradientBoostingRegressor': {'score': 0.9550335570469798,
  'model': GradientBoostingClassifier()},
 'XGBoost': {'score': 0.9436241610738255, 'model': XGBClassifier()},
 'BaggingClassifier': {'score': 0.8684563758389261,
  'model': BaggingClassifier()},
 'MultinomialNB': {'score': 0.9724832214765101, 'model': MultinomialNB()}}

We can see that the Accuracy value is greater for Logistic regression followed by Naive bayers model

LogisticRegression': 'score': 0.9751677852348992

'MultinomialNB': 'score': 0.9724832214765101

GradientBoostingRegressor': 'score': 0.9550335570469798

RandomForestClassifier': 'score': 0.9516778523489933

'XGBoost': 'score': 0.9436241610738255

'BaggingClassifier': 'score': 0.8684563758389261

